### ISSUE TITLE
[mlir] Move SPIR-V narrow bitwidth storage emulation to MemRef

### ISSUE THREAD

**Issue by antiagainst** [llvm/llvm-project#62662]
We have quite some logic in [MemRef To SPIR-V conversions](https://github.com/llvm/llvm-project/blob/0c7f3d6/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp) that emulates narrow bitwidth storage support if they are missing in the target environment, like supporting i8 load/store via i32. They are added way earlier when MLIR dialects were not as developed and structured as today. Such logic won't necessary need to couple with SPIR-V; we can use the same mechanism for conversion to LLVM. We should extract them out into a MemRef level conversion for better structure and broader usage. This will also be a nice counterpart to [`EmulateWideInt`](https://github.com/llvm/llvm-project/blob/faafd26/mlir/lib/Dialect/MemRef/Transforms/EmulateWideInt.cpp) we already have there. 

For now we only need to worry about storage; not compute emulation. Specifically, we need to

* Add a MemRef type converter to convert `memref` of narrow bitwidth (e.g., i8) into wider bitwidth (e.g., i32),
* Convert allocation/load/store op size/index and do bit manipulation to get the proper narrow bitwidth value. 

The type converter should be configurable regarding what integer types are directly allowed and what integer types we should use for emulation, like EmulateWideInt. 

Such logic already exists in MemRef to SPIR-V conversions; we just need to generate `arith` ops for various caculations instead of SPIR-V ones.

**Comment by antiagainst** [comment]
@yzhang93 FYI

**Comment by MaheshRavishankar** [comment]
One issue to address here is that the SPIR-V version of the emulation here works on 1D memref types. To make this work on N-D memrefs, we need to use the [`memref.extract_strided_metadata`](https://github.com/llvm/llvm-project/blob/0245dc94ebee82a14118cdb1b39c9f4470e99e1d/mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td#L900) operation. This extracts the `base_ptr`, `offset`, `sizes` and `strides` for the value of `memref` types.

```
%0 = memref.load %0[%v0][%v1] : memref<?x?xi4, strided<[?, ?], offset = ?>>
```

can be replaced with 

```
%b, %offset, %sizes:2, %strides:2 = memref.extract_strided_metadata %0
%linearized_offset = /// %v0 * %stride#0 + %v1 * %stride#1
%linearized_size = /// %size0 * %size1
%linearized = memref.reinterpret_cast %b, offset = [%offset], sizes = [%linearized_size], strides = [%stride#1] 
%load = memref.load %linearized[%linearized_offset] : memref<?xi4, strided<?, offset = ?>>
```

Then the logic for handling the offsets will work for multi-dimensional memrefs.

**Comment by hanhanW** [comment]
Thanks Lei and Mahesh for the pointers. Below is my findings about emulating subtypes in LLVM.

On SPIR-V, we can emulate narrow int types load/store:
1. Compute the base pointer linearized indices
2. Extract the element. It loads/stores multiple elements at once.
3. For load case, we [shift and truncate](https://github.com/llvm/llvm-project/blob/0c7f3d6c39d7dd7d240854ac347a801d62dc938b/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp#L469-L490) the data to get the actual value.
4. For store case, we'll need to an atomic operation to store the value because of race conditions in multi-threads. On SPIR-V side, we use [AtomicAnd to clear the destination data and AtomicOr to store the value](https://github.com/llvm/llvm-project/blob/0c7f3d6c39d7dd7d240854ac347a801d62dc938b/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp#L590-L627).

On the LLVM side, I've found some equivalent methods and places to emulate narrow int cases for short-term solution. To have similar things, we'll need
1. Extend [LLVM type converter](https://github.com/llvm/llvm-project/blob/f649599ea93301bd0d0a2b8e450d1f77425ea92e/mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp#LL36C48-L36C66) to accept int4 cases, like [SPIRVTypeConverter](https://github.com/llvm/llvm-project/blob/7d2b7be28d903a74f3ca10b2d5f97d69e006df62/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L585-L591). We probably have to add some options like `LLVMSubByteTypeStorage ::Pack`, so it does not simply break downstream projects. Here is the [implementation](https://github.com/llvm/llvm-project/blob/f649599ea93301bd0d0a2b8e450d1f77425ea92e/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L268-L271) on SPIR-V side.
2. Adapt the [adjustment methods](https://github.com/llvm/llvm-project/blob/0c7f3d6c39d7dd7d240854ac347a801d62dc938b/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp#L31-L90) using LLVM dialect operations. The LLVM side has [getStridedElementPtr method](https://github.com/llvm/llvm-project/blob/71703a097859a24883aa32c3ee258647412c311e/mlir/lib/Conversion/LLVMCommon/Pattern.cpp#L68), which is similar to [spirv::getElementPtr method](https://github.com/llvm/llvm-project/blob/5a7563b2f675f76e6c4e1e955610f695b7bdbfc8/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L938-L950) on SPIR-V side. We probably need to modify the method based on needs.
3. Add a pattern to emulate int4/int16 loads to [memref.load -> LLVMIR](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/MemRefToLLVM/MemRefToLLVM.cpp) like SPIR-V.
4. Add a pattern to emulate int4/int16 stores to [memref.store -> LLVMIR](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/MemRefToLLVM/MemRefToLLVM.cpp). SPIT-V emulates it using two atomic operations. On LLVM side, we can probably use [AtomicRMWOp](https://github.com/llvm/llvm-project/blob/0245dc94ebee82a14118cdb1b39c9f4470e99e1d/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L1765) for handling it.

Once we have things in place, we are able to emulate narrow ints on LLVM like SPIR-V.

For long-term solution, we will need a `EmulateNarrowInt` pass, which is a counterpart of [EmulateWideInt](https://github.com/llvm/llvm-project/blob/faafd26/mlir/lib/Dialect/MemRef/Transforms/EmulateWideInt.cpp). What we need to do is

1. Implement `ConvertMemRefAlloc`, `ConvertMemRefLoad`, and `ConvertMemRefStore` for narrow int cases. We can use memref::AtomicRMWOp for store case to avoid race condition issue. We'll need `memref.extract_strided_metadata` to help compute base pointer and offsets, like Mahesh mentioned above.
2. Implement bunch of patterns to replace arith operations with new types. There are a couple of categories:
  - Casting ops. We will replace op with new types if types mismatch; replace the op with the source if the type does not change.
  - Binary ops. They can be covered by a single template pattern.
  - Unary ops. Simliar to binary ops.
3. Once we have above supports, we can create a LLVMEmulateNarrowInt pass and use it in LLVMConversion.
4. We'll be able to move SPIR-V to reuse those patterns with different type converter.

I think @yzhang93 and I can go with the second solution directly. @antiagainst , does it align with what you have in mind?

**Comment by yzhang93** [comment]
@hanhanW Thanks for the details. I'm going to read through the relevant codes. Just want to confirm, we should directly aim for the long-term solution, right? And from our last meeting, Mahesh mentioned I should first have the load part functional and the CPU tests can be unblocked. It seems store need more work, so shall we split this into two separate stages/patches?

**Comment by MaheshRavishankar** [comment]
> Thanks Lei and Mahesh for the pointers. Below is my findings about emulating subtypes in LLVM.
> 
> On SPIR-V, we can emulate narrow int types load/store:
> 
> 1. Compute the base pointer linearized indices
> 2. Extract the element. It loads/stores multiple elements at once.
> 3. For load case, we [shift and truncate](https://github.com/llvm/llvm-project/blob/0c7f3d6c39d7dd7d240854ac347a801d62dc938b/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp#L469-L490) the data to get the actual value.
> 4. For store case, we'll need to an atomic operation to store the value because of race conditions in multi-threads. On SPIR-V side, we use [AtomicAnd to clear the destination data and AtomicOr to store the value](https://github.com/llvm/llvm-project/blob/0c7f3d6c39d7dd7d240854ac347a801d62dc938b/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp#L590-L627).
> 
> On the LLVM side, I've found some equivalent methods and places to emulate narrow int cases for short-term solution. To have similar things, we'll need
> 
> 1. Extend [LLVM type converter](https://github.com/llvm/llvm-project/blob/f649599ea93301bd0d0a2b8e450d1f77425ea92e/mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp#LL36C48-L36C66) to accept int4 cases, like [SPIRVTypeConverter](https://github.com/llvm/llvm-project/blob/7d2b7be28d903a74f3ca10b2d5f97d69e006df62/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L585-L591). We probably have to add some options like `LLVMSubByteTypeStorage ::Pack`, so it does not simply break downstream projects. Here is the [implementation](https://github.com/llvm/llvm-project/blob/f649599ea93301bd0d0a2b8e450d1f77425ea92e/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L268-L271) on SPIR-V side.
> 2. Adapt the [adjustment methods](https://github.com/llvm/llvm-project/blob/0c7f3d6c39d7dd7d240854ac347a801d62dc938b/mlir/lib/Conversion/MemRefToSPIRV/MemRefToSPIRV.cpp#L31-L90) using LLVM dialect operations. The LLVM side has [getStridedElementPtr method](https://github.com/llvm/llvm-project/blob/71703a097859a24883aa32c3ee258647412c311e/mlir/lib/Conversion/LLVMCommon/Pattern.cpp#L68), which is similar to [spirv::getElementPtr method](https://github.com/llvm/llvm-project/blob/5a7563b2f675f76e6c4e1e955610f695b7bdbfc8/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L938-L950) on SPIR-V side. We probably need to modify the method based on needs.
> 3. Add a pattern to emulate int4/int16 loads to [memref.load -> LLVMIR](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/MemRefToLLVM/MemRefToLLVM.cpp?rgh-link-date=2023-05-16T00%3A02%3A00Z) like SPIR-V.
> 4. Add a pattern to emulate int4/int16 stores to [memref.store -> LLVMIR](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/MemRefToLLVM/MemRefToLLVM.cpp?rgh-link-date=2023-05-16T00%3A02%3A00Z). SPIT-V emulates it using two atomic operations. On LLVM side, we can probably use [AtomicRMWOp](https://github.com/llvm/llvm-project/blob/0245dc94ebee82a14118cdb1b39c9f4470e99e1d/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L1765) for handling it.
> 
> Once we have things in place, we are able to emulate narrow ints on LLVM like SPIR-V.

I am not sure we want to go down path of doing things in LLVMTypeConverter. Its too opaque, and will be superceded by doing this in `memref` anyway....

> 
> For long-term solution, we will need a `EmulateNarrowInt` pass, which is a counterpart of [EmulateWideInt](https://github.com/llvm/llvm-project/blob/faafd26/mlir/lib/Dialect/MemRef/Transforms/EmulateWideInt.cpp?rgh-link-date=2023-05-16T00%3A02%3A00Z). What we need to do is
> 
> 1. Implement `ConvertMemRefAlloc`, `ConvertMemRefLoad`, and `ConvertMemRefStore` for narrow int cases. We can use memref::AtomicRMWOp for store case to avoid race condition issue. We'll need `memref.extract_strided_metadata` to help compute base pointer and offsets, like Mahesh mentioned above.
> 2. Implement bunch of patterns to replace arith operations with new types. There are a couple of categories:
> 
> * Casting ops. We will replace op with new types if types mismatch; replace the op with the source if the type does not change.
> * Binary ops. They can be covered by a single template pattern.
> * Unary ops. Simliar to binary ops.
> 
> 3. Once we have above supports, we can create a LLVMEmulateNarrowInt pass and use it in LLVMConversion.
> 4. We'll be able to move SPIR-V to reuse those patterns with different type converter.
> 
> I think @yzhang93 and I can go with the second solution directly. @antiagainst , does it align with what you have in mind?

The long terms solution makes sense to me.

**Comment by hanhanW** [comment]
Yes, it's better to start with the long-term solution; we can add patterns incrementally. In the first patch, let's start with load op. I'll think more about it and circle back.

**Comment by antiagainst** [comment]
+1 towards doing it directly at MemRef level and starting with load ops.

**Comment by yzhang93** [comment]
@hanhanW I'm still digesting all different parts, and do have some questions. It would be really helpful if we can setup a VC and go through them. I'll ping you through google chat. Thanks!

**Comment by hanhanW** [comment]
We have load patterns working, and @saienduri is working on store part. Emulation for vector.store is landed in MLIR, and we will verify correctness in IREE (https://github.com/openxla/iree/issues/15369). The next item is memref.store (https://github.com/openxla/iree/issues/15370).
